### DISCUSSION TITLE
Difference in optimization between Linq's OfType<T> and Cast<T>

### DISCUSSION THREAD

**Discussion by IAmTheCShark** [dotnet/runtime#54496] [Q&A] [answered]
I know what theese methods do under the hood.
What I do not understand here is the fact that `Cast<T>` has an optimization that checks if the IEnumerable is an` IEnumerable<T>`.
`OfType<T>` doesnt do this.

My question is why?

Problem is, that those "redundant" OfType-calls wrap every IEnumerable in a `OfTypeIterator`  which causes optimizations like the ICollection thing in methods like  `Count()` to not work.

**Comment by yaakov-h** [comment, marked as answer]
This is a question for https://github.com/dotnet/runtime.

**Reply by IAmTheCShark** [reply]
Whoops, my bad. Ill transfer it there.
https://github.com/dotnet/runtime/discussions/54473
